### DISCUSSION TITLE
Default "serverpod create" project tries to connect to the wrong port

### DISCUSSION THREAD

**Discussion by lukehutch** [serverpod/serverpod#521] [Q&A] [answered]
I tried creating a default project using `serverpod create <podname>`, then started the server in docker (it starts on port 8080), then ran the flutter code using `flutter run`.

The app fails when you try to send the inputted name to the server, with:

```
flutter: Failed call: example.hello
flutter: SocketException: Connection refused (OS Error: Connection refused, errno = 111),
        address = localhost, port = 56396
```

The Flutter code tries to connect to port 56396 (or other random port). This happens despite the code stating it is supposed to be connecting to port 8080:

```
var client = Client('http://localhost:8080/')
  ..connectivityMonitor = FlutterConnectivityMonitor();
```

**Comment by yahu1031** [comment]
@lukehutch I have faced a similar issue. But hope you aren't running this on the mobile platform. If it is in mac, then you need to allow client connection in `x_flutter\macos\Runner\DebugProfile.entitlements` and also `x_flutter\macos\Runner\Release.entitlements`.

```diff
<plist version="1.0">
<dict>

	<key>com.apple.security.app-sandbox</key>
	<true/>
	...
++      <key>com.apple.security.network.client</key>
++      <true/>

</dict>
</plist>
```

**Comment by lukehutch** [comment]
I tried running both the Linux client and the Web client -- both have the same problem.

The server is not OS-specific, I think?

**Comment by vlidholt** [comment, marked as answer]
@lukehutch Have you successfully started the server by running `dart bin/main.dart` in the server directory?

The way the socket works when it opens a connection to a server is that it also opens a new port for the client, which it uses to receive data. That port will be opened on a random port number and is what you see in your logs. It still tries to connect to the correct port.

**Comment by lukehutch** [comment]
@vlidholt Ah, no, sorry, I missed that line in the instructions.

After running `dart bin/main.dart` I get errors of the form:

```
2022-12-29 09:17:09.744915Z Internal server error. Failed to connect to database in future call manager.
PostgreSQLSeverity.fatal 28P01: password authentication failed for user "postgres"
```

**Reply by yahu1031** [reply]
This is because Your password was wrong when you are trying to connect. Please check that with any Postgres DB GUI tools. If it connects there then it must actually work.

**Reply by lukehutch** [reply]
I'm having trouble connecting to the Postgres instance in the docker container. It doesn't seem to map to localhost:5432. How am I supposed to interact with the dockerized Postgres, using say pgadmin?

(The instructions on the example code need to be udpated to tell new users how to do this -- they just say to set up the database tables using the generated SQL instructions. I forgot that I had skipped that step too, because I couldn't figure out how to connect to Postgres.)

**Reply by RohKum** [reply]
> @vlidholt Ah, no, sorry, I missed that line in the instructions.
> 
> After running `dart bin/main.dart` I get errors of the form:
> 
> ```
> 2022-12-29 09:17:09.744915Z Internal server error. Failed to connect to database in future call manager.
> PostgreSQLSeverity.fatal 28P01: password authentication failed for user "postgres"
> ```

The Solution to the Problem
Checkout this link : - https://stackoverflow.com/questions/76172995/serverpod-unable-to-authenticate-with-postgres

**Comment by lukehutch** [comment]
There seems to be an issue with `serverpod create` that causes postgres to be set up wrongly about 80% of the time.

https://github.com/serverpod/serverpod/issues/531

**Comment by lukehutch** [comment]
So if I run `serverpod create` enough times, I get a test pod that works.

However, I can't figure out how to connect `pgadmin` to the Postgres installation in the Docker instance. How do I do that?

`ps aux` gives the following line:

```
systemd+ 2917887  0.0  0.0 214928 16388 ?        Ss   17:18   0:00 postgres: postgres testpod 172.22.0.1(51048) idle
```

I assume the Postgres instance is exposed on the Docker IP address 172.22.0.1, at port 51048? But pgadmin can't connect to that.

How does `testpod_server` even find the Docker instance's IP address and the Postgres port?

**Reply by vlidholt** [reply]
By default, Serverpod's Postgres instance will run on port 8090. You can find all the details for connecting in the `config/development.yaml` file and in the `config/passwords.yaml` file.

**Reply by lukehutch** [reply]
@vlidholt Thank you, that was the information I needed. I can now connect to the database from pgAdmin. Can you please add that info to the documentation?

**Reply by vlidholt** [reply]
Will do! Thank you!

**Reply by Bierbuddy** [reply]
Can you please explain to me how you solved that issue. I still have the same problem and i invested the day and it still doesn't work

**Reply by lukehutch** [reply]
@Bierbuddy follow the instructions of @vlidholt above -- go look in `config/development.yaml` and `config/passwords.yaml` -- that gives you all the information you need to connect to your Postgres instance, using any database frontend. You also need to use the same hostname and port when you create the `Client`. See https://github.com/serverpod/serverpod/issues/544#issuecomment-1376339244

**Reply by Bierbuddy** [reply]
Sorry that is to high for me. I dont understand what to do in these files. I tried the steps presented here on stackoverflow. But it still doesn't work.

https://stackoverflow.com/questions/76180598/serverpod-sql-error-when-starting-a-clean-project/76180599#76180599

Is there any possibility to uninstall and reinstall everything so it works? Or that i change the emulator? Or is there any YouTube-Tutorial that explains the process?

**Reply by lukehutch** [reply]
@vlidholt can you guys have `serverpod create` ask the user some basic questions like hostname, port, whether the user will be using the Android emulator, etc. -- so that all this can be set up correctly right from the beginning? It took me a while to struggle through all that too.

Also, there should be a guide in the docs (if it's not already there) for everything that needs to be changed when moving from development on localhost to a hosted solution.

**Reply by SandPod** [reply]
@lukehutch (interactive `serverpod create`) Good suggestion, we have talked about it a bit internally. Will draft up a feature request for it so that we can track the discussion there.

Will also add a documentation issues to make sure we create a guide for how to move from localhost to a hosted solution. I think the information required today is spread out in different sections in our documentation, but It would be good to collect it to a single guide.

**Reply by SandPod** [reply]
Issue for interactive mode: https://github.com/serverpod/serverpod/issues/1643
Issue for documenting how to move to a hosted solution: https://github.com/serverpod/serverpod/issues/1644
